Reduce special treatment for zsts

oli-obk opened the pull request:
addresses https://github.com/rust-lang/rust/pull/67467#discussion_r360650846

cc @RalfJung

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

oli-obk commented:
r? @RalfJung

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=16437&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16437/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16437/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-21T23:39:06.3234542Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-21T23:39:06.3413682Z ##[command]git config gc.auto 0
2019-12-21T23:39:06.3511171Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-21T23:39:06.3536184Z ##[command]git config --get-all http.proxy
2019-12-21T23:39:06.3721780Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/67501/merge:refs/remotes/pull/67501/merge
---
2019-12-21T23:44:46.1061649Z    Compiling serde_json v1.0.40
2019-12-21T23:44:47.6543766Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2019-12-21T23:44:57.7308178Z     Finished release [optimized] target(s) in 1m 19s
2019-12-21T23:44:57.7400879Z tidy check
2019-12-21T23:44:58.5787293Z tidy error: /checkout/src/librustc_mir/interpret/place.rs:238: line longer than 100 chars
2019-12-21T23:44:58.5788009Z tidy error: /checkout/src/librustc_mir/interpret/place.rs:241: line longer than 100 chars
2019-12-21T23:45:00.1559536Z Found 485 error codes
2019-12-21T23:45:00.1559660Z Found 0 error codes with no tests
2019-12-21T23:45:00.1559727Z Done!
2019-12-21T23:45:00.1559769Z some tidy checks failed
2019-12-21T23:45:00.1559769Z some tidy checks failed
2019-12-21T23:45:00.1559796Z 
2019-12-21T23:45:00.1559823Z 
2019-12-21T23:45:00.1561413Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor"
2019-12-21T23:45:00.1561607Z 
2019-12-21T23:45:00.1561664Z 
2019-12-21T23:45:00.1568203Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2019-12-21T23:45:00.1568266Z Build completed unsuccessfully in 0:01:23
2019-12-21T23:45:00.1568266Z Build completed unsuccessfully in 0:01:23
2019-12-21T23:45:00.1623415Z == clock drift check ==
2019-12-21T23:45:00.1636536Z   local time: Sat Dec 21 23:45:00 UTC 2019
2019-12-21T23:45:00.3132323Z   network time: Sat, 21 Dec 2019 23:45:00 GMT
2019-12-21T23:45:01.0447398Z == end clock drift check ==
2019-12-21T23:45:01.6958084Z 
2019-12-21T23:45:01.7022412Z ##[error]Bash exited with code '1'.
2019-12-21T23:45:01.7051415Z ##[section]Starting: Checkout
2019-12-21T23:45:01.7053125Z ==============================================================================
2019-12-21T23:45:01.7053202Z Task         : Get sources
2019-12-21T23:45:01.7053253Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #67540) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #67596) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -216,21 +216,40 @@ impl<'tcx, Tag> MPlaceTy<'tcx, Tag> {
     }
+
+    #[inline(always)]
+    pub fn to_const_value(self, tcx: TyCtxt<'tcx>) -> ConstValue<'tcx> {
```
This is an odd function... why is it here, and not in `const_eval.rs` where we have most of the other conversion functions?

RalfJung replied:
Also, converting a *place* to a *value* doesn't make much sense to begin with. The returned value represents the content of the place, right (the "pointee")?

oli-obk replied:
true. I could name this function better.

I'll also move it.

oli-obk replied:
it's kind of a "read". But since a `Place` is just an "rvalue reference" in c++ terms, converting it to a value seemed ok to me.

I inlined the function as a closure at the use site

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -216,21 +216,40 @@ impl<'tcx, Tag> MPlaceTy<'tcx, Tag> {
+            },
+            Scalar::Raw { data, .. } => {
+                assert_eq!(data, self.layout.align.abi.bytes().into());
+                ConstValue::Scalar(Scalar::zst())
```
Why does this make any sense?

oli-obk replied:
If the mplace has an integer address and you read from it at compile time (which is what all users of this function want), it must be a ZST.

RalfJung replied:
Makes sense, but it'd be better if we could assert this. Can't we do `assert!(self.layout.zst())`?

RalfJung commented:
This is quite clever! I am conflicted. I don't like the weird condition in `try_as_mplace` very much, and I don't like how this spreads values obtained by `MemPlace::dangling`, which I don't feel should even existing in the spec so I wonder why we have a hard time avoiding it here. On the other hand, I do like the simplifications this brings.

The previous uses of `dangling`, I think, were never actually observable by the program. Basically, what we should never do is call `dangling` and then later call `to_ref`. Could you add that in a doc-comment, and at least superficially audit the users of `try_as_mplace` accordingly?

Do you think there is any chance to also kill the other use of `MemPlace::dangling`, and then remove the `dangling` function entirely and inline it into `try_as_mplace`?

oli-obk commented:
There are also two uses of `dangling` in miri for uses of `call_function` where the invoked function is statically known to have unit return value.

oli-obk commented:
Even if we removed `dangling`, you could achieve the same effect via

```rust
OpTy::from(ImmTy::from(Scalar::zst(), layout)).try_as_mplace()
```

So I'm not sure what we'd have gained.

RalfJung commented:
> There are also two uses of dangling in miri for uses of call_function where the invoked function is statically known to have unit return value.

Two? I remember one, for calling `drop`. But yeah, there's probably no good replacement for that.

This point still stands, though:

> The previous uses of dangling, I think, were never actually observable by the program. Basically, what we should never do is call dangling and then later call to_ref. Could you add that in a doc-comment, and at least superficially audit the users of try_as_mplace accordingly?

oli-obk commented:
> Two? I remember one, for calling drop. But yeah, there's probably no good replacement for that.

Two in miri, and one in the miri engine (the one for drop)

oli-obk commented:
> Basically, what we should never do is call dangling and then later call to_ref.

That feels impossible to uphold and I don't see what we'd gain. Do you want to prevent `(&() as *const () as usize) / 2` to fail because the value is a pointer but `(&*&() as *const () as usize) / 2` to succeed? Because this is I think what would happen with the current PR with the appropriate feature gates.

RalfJung commented:
We'd gain not baking these fake dangling points into the implemented semantics of the Abstract Machine in an observable way.  Or rather, that's a property that we currently have, and I do not want to lose it.

RalfJung commented:
(Dang, looks like GH stopped sending email notifications for force-pushes. So please let me know when this is ready for review again.)

oli-obk commented:
> We'd gain not baking these fake dangling points into the implemented semantics of the Abstract Machine in an observable way.

I can poison the `MemPlace` when it is created via `dangling` and make `as_ref` panic if the `MemPlace` is poisoned.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=16891&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16891/logs/54), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16891/logs/54)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-26T22:55:28.0677629Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-26T22:55:28.0877646Z ##[command]git config gc.auto 0
2019-12-26T22:55:28.0958304Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-26T22:55:28.1003323Z ##[command]git config --get-all http.proxy
2019-12-26T22:55:28.1157041Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/67501/merge:refs/remotes/pull/67501/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented:
I addressed all the review comments.

> at least superficially audit the users of try_as_mplace accordingly?

I did the audit and found no problematic cases. All cases are either `Operand` uses, so we'll definitely read later or write to the memory, but want to have the address back for anything other than getting the backing memory.

bors commented:
:umbrella: The latest upstream changes (presumably #67192) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented on src/librustc_mir/const_eval/eval_queries.rs:
```diff
@@ -115,28 +115,26 @@ pub(super) fn op_to_const<'tcx>(
             ConstValue::ByRef { alloc, offset: ptr.offset }
         }
+        Scalar::Raw { data, .. } => {
+            assert_eq!(data, mplace.layout.align.abi.bytes().into());
```
Please add a comment to this assertion... I get it now but it's still funny.

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -20,6 +20,45 @@ use super::{
 
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, HashStable)]
+pub enum MemPlaceMeta<Tag = (), Id = AllocId> {
+    Unsized(Scalar<Tag, Id>),
```
Please also add a doc comment to this variant, and to the type itself.

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -20,6 +20,45 @@ use super::{
+    /// a ZST Operand with a backing allocation and being converted to an integer address. This
+    /// should be impossible, because you can't take the address of an operand, but this is a second
+    /// protection layer ensuring that we don't mess up.
+    Poison,
```
Am I missing something or did you forget to actually use this variant for `MPlaceTy::dangling`?

oli-obk replied:
:man_facepalming:

RalfJung commented:
All right, with that extra dynamic check I this the pros definitely outweigh the cons for this approach. Thanks!

joelpalmer commented:
Ping from Triage: any updates @oli-obk?

RalfJung commented on src/librustc_mir/const_eval/eval_queries.rs:
```diff
@@ -115,28 +115,31 @@ pub(super) fn op_to_const<'tcx>(
+                "this MPlaceTy must come from `try_as_mplace` being used on a zst, so we know what
+                 value this integer address must have",
+            );
+            assert!(mplace.layout.is_zst());
```
Nit: I think it makes sense to check this first.

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -139,13 +180,9 @@ impl<'tcx, Tag> MPlaceTy<'tcx, Tag> {
-        }
+        let align = layout.align.abi;
+        let ptr = Scalar::from_uint(align.bytes(), cx.pointer_size());
+        MPlaceTy { mplace: MemPlace { ptr, align, meta: MemPlaceMeta::Poison }, layout }
```
```suggestion
        // `Poison` this to make sure that the pointer value `ptr` is never observable by the program.
        MPlaceTy { mplace: MemPlace { ptr, align, meta: MemPlaceMeta::Poison }, layout }
```

RalfJung commented on src/librustc_mir/interpret/place.rs:
```diff
@@ -20,6 +20,47 @@ use super::{
+/// Information required for the sound usage of a `MemPlace`.
+pub enum MemPlaceMeta<Tag = (), Id = AllocId> {
+    /// The unsized payload (e.g. length for slices or vtable pointer for trait objects).
+    Unsized(Scalar<Tag, Id>),
```
There is a potential confusion here because some unsized types don't have metadata -- namely, `extern type`. Maybe `DynSized` would be a better name?

oli-obk replied:
I don't know about any difference between the meaning of *unsized* and *dynsized*.

Doesn't `extern type` also have metdata but it's a zst?

RalfJung replied:
Dynsized means that dynamically, we can determine the size. That's true for `dyn Trait` but not for `extern type`.

> Doesn't extern type also have metdata but it's a zst?

Well... does it? I might as well argue that `&i32` has metadata but it's a ZST.

oli-obk replied:
> That's true for dyn Trait but not for extern type.

Well, it's still up for discussion whether `size_of_val` should work for extern types. Maybe let's stay with `Unsized` and mention `extern type` in the docs?

RalfJung replied:
The discussion is only whether it should panic or be prevent to be called through some kind of bound, isn't it? AFAIK the proposal to return a bogus value never got real traction (and rightly so).

oli-obk replied:
no, not bogus values. What i mean is that e.g. `CStr` can't become an extern type without us adding some scheme that allows extern types to work with `size_of_val`. Right now we have `struct CStr([u8]);` which works as expected, but if we moved to `struct CStr(ExternCStr);` we'd be doing a breaking change because `size_of_val` would now panic.

RalfJung replied:
That sounds to me like the plans for custom DST (dynamically sized types). That's far off, but I see.

Maybe just call the variant `Meta` or so then?

RalfJung commented:
Just a couple nits. r=me with those fixed.

oli-obk commented:
@bors r=RalfJung

bors commented:
:pushpin: Commit c5c4fa8e7633c28464e3b27a57f2f175cc6700fd has been approved by `RalfJung`

<!-- @bors r=RalfJung c5c4fa8e7633c28464e3b27a57f2f175cc6700fd -->
<!-- homu: {"type":"Approved","sha":"c5c4fa8e7633c28464e3b27a57f2f175cc6700fd","approver":"RalfJung"} -->

RalfJung commented:
@oli-obk Miri broke in a rollup containing this PR, and this seems like the likely culprit. Could you prepare a fixup PR?

oli-obk commented:
on it
